addons/account/models/company.py

```diff
@@ -414,36 +467,143 @@ def _validate_fiscalyear_lock(self, values):
                 raise RedirectWarning(msg, action, _('Show Journal Entries to Hash'))
 
-    def _get_user_fiscal_lock_date(self):
-        """Get the fiscal lock date for this company depending on the user"""
-        lock_date = max(self.period_lock_date or date.min, self.fiscalyear_lock_date or date.min)
-        if self.env.user.has_group('account.group_account_manager'):
-            lock_date = self.fiscalyear_lock_date or date.min
-        if self.parent_id:
-            # We need to use sudo, since we might not have access to a parent company.
-            lock_date = max(lock_date, self.sudo().parent_id._get_user_fiscal_lock_date())
-        return lock_date
-
-    def _get_violated_lock_dates(self, accounting_date, has_tax):
+    def _get_user_lock_date(self, soft_lock_date_field, ignore_exceptions=False):
+        """Get the lock date called `soft_lock_date_field` for this company depending on the user.
+        We consider the field and exceptions (except if `ignore_exceptions`) for it in this company and the parent companies.
+        :param str soft_lock_date_field: One of the lock date fields (except 'hard_lock_date'; see SOFT_LOCK_DATE_FIELDS)
+        :param bool ignore_exceptions: Whether we ignore exceptions or not
+        :return the user lock date
+        """
+        self.ensure_one()
+        soft_lock_date = date.min
+        # We need to use sudo, since we might not have access to a parent company.
+        for company in self.sudo().parent_ids:
+            if company[soft_lock_date_field]:
+                if ignore_exceptions:
+                    exception = None
+                else:
+                    exception = self.env['account.lock_exception'].search(
+                        [
+                          ('state', '=', 'active'),  # checks the datetime
+                          '|',
+                              ('user_id', '=', None),
+                              ('user_id', '=', self.env.user.id),
+                          (soft_lock_date_field, '<', company[soft_lock_date_field]),
+                          ('company_id', '=', company.id),
+                        ],
+                        order=f'{soft_lock_date_field} asc',
+                        limit=1,
+                    )
+                if exception:
+                    soft_lock_date = max(soft_lock_date, exception[soft_lock_date_field])
+                else:
+                    soft_lock_date = max(soft_lock_date, company[soft_lock_date_field])
+        return soft_lock_date
+
+    def _get_user_fiscal_lock_date(self, journal, ignore_exceptions=False):
+        """Get the fiscal lock date for this company (depending on the affected journal) accounting for potential user exceptions
+        :param bool ignore_exceptions: Whether we ignore exceptions or not
+        :return the lock date
+        """
+        self.ensure_one()
+        user = self.env.user.with_company(self).with_context(ignore_exceptions=ignore_exceptions)
```

Comment:
Good points. I moved the fields.
As discussed: We need the context key though (or an additional field) since 1 field needs to store the actual value.